src/Struct.php

```diff
@@ -0,0 +1,223 @@
+ * invalid PHP syntax, use curly braces around a quoted property name:
+ *
+ * Example: `$struct->{"bob page"}`
+ *
+ * Lastly, this class only reserves one property name: "__data". The property is prefixed with a double underscore to
+ * reduce the probability of conflicting with any consumer property name. However, it is recommended to avoid to using
+ * double underscore prefixes as additional properties may be added in future versions to accommodate new features.
+ *
+ * @since [*next-version*]
+ */
+abstract class Struct implements Serializable
+{
+    /**
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $__data;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $data A mapping of property names to their values.
+     */
+    public function __construct(array $data = [])
+    {
+        $props = $this->getPropTypes();
+
+        $this->__data = [];
+        foreach ($props as $key => $type) {
+            $this->__data[$key] = isset($data[$key])
+                ? $type->cast($data[$key])
+                : $type->getDefault();
+
+            unset($data[$key]);
+        }
+
+        if (!empty($data)) {
+            $class = static::class;
+            $props = implode(', ', array_keys($data));
+            throw new LogicException("Struct {$class} does not have the following properties: {$props}");
+        }
+    }
+
+    /**
+     * Retrieves the value for a property.
+     *
+     * @since [*next-version*]
+     *
+     * @param string $name The name of the property.
+     *
+     * @return mixed|null The value of the property.
+     */
+    public function __get($name)
+    {
+        $props = $this->getPropTypes();
+
+        if (!array_key_exists($name, $props)) {
+            throw static::createUndefinedPropException($name);
+        }
+
+        return $this->__data[$name];
+    }
+
+    /**
+     * Should not be used. Only exists to warn developers of misuse.
+     *
+     * @since [*next-version*]
+     *
+     * @param string $name  The name of the property.
+     * @param mixed  $value The value being assigned to the property.
+     *
+     * @throws Exception Always throws.
+     */
+    public function __set($name, $value)
+    {
+        throw new LogicException('Cannot set property values. Use with() instead');
+    }
+
+    /**
+     * Creates a copy of the struct instance with some modified properties.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $changes An associative array that maps property names to their new values.
+     *
+     * @return Struct The copied struct with the applied changes. If the $changes parameter is empty, no copy is
+     *                performed and the same instance is returned.
+     */
+    public function with(array $changes)
+    {
+        if (empty($changes)) {
+            return $this;
```

Comment:
It would be IMHO more intuitive if here you'd return the `$clone` instead.

Reply:
Well, there is functionally no difference. Since a struct is immutable, instance does not matter. That's like saying you prefer this `5` over another `5`. The number `5` is always `5`, and has its own identity.

Similarly, here it doesn't matter if the instance is the same one or a clone with no changes. The struct has its own identity, which is its internal data. So to preserve memory and time, I return `$this`.